File: code/data_structures/balanced_binary_search_tree/AVLTreeRecursiveOptimized.cpp
#include <iostream>
#include <stack>
#include <stdexcept>

template <typename T>
class AVLTreeRecursiveOptimized {
private:
    struct Node {
        int bf;
        T value;
        int height;
        Node* left;
        Node* right;

        Node(const T& val) : value(val), bf(0), height(0), left(nullptr), right(nullptr) {}
    };

    Node* root;
    int nodeCount;
    Node* TOKEN;

    Node* insert(Node* node, const T& value) {
        if (node == nullptr) {
            nodeCount++;
            return new Node(value);
        }

        int cmp = value - node->value;

        if (cmp < 0) {
            Node* newLeftNode = insert(node->left, value);
            if (newLeftNode == TOKEN) return TOKEN;
            node->left = newLeftNode;
        } else if (cmp > 0) {
            Node* newRightNode = insert(node->right, value);
            if (newRightNode == TOKEN) return TOKEN;
            node->right = newRightNode;
        } else {
            return TOKEN;
        }

        update(node);
        return balance(node);
    }

    void update(Node* node) {
        int leftNodeHeight = (node->left == nullptr) ? -1 : node->left->height;
        int rightNodeHeight = (node->right == nullptr) ? -1 : node->right->height;

        node->height = 1 + std::max(leftNodeHeight, rightNodeHeight);
        node->bf = rightNodeHeight - leftNodeHeight;
    }

    Node* balance(Node* node) {
        if (node->bf == -2) {
            if (node->left->bf <= 0) {
                return rightRotation(node);
            } else {
                return leftRightCase(node);
            }
        } else if (node->bf == 2) {
            if (node->right->bf >= 0) {
                return leftRotation(node);
            } else {
                return rightLeftCase(node);
            }
        }
        return node;
    }

    Node* rightRotation(Node* node) {
        Node* newParent = node->left;
        node->left = newParent->right;
        newParent->right = node;
        update(node);
        update(newParent);
        return newParent;
    }

    Node* leftRotation(Node* node) {
        Node* newParent = node->right;
        node->right = newParent->left;
        newParent->left = node;
        update(node);
        update(newParent);
        return newParent;
    }

    Node* leftLeftCase(Node* node) {
        return rightRotation(node);
    }

    Node* leftRightCase(Node* node) {
        node->left = leftRotation(node->left);
        return leftLeftCase(node);
    }

    Node* rightRightCase(Node* node) {
        return leftRotation(node);
    }

    Node* rightLeftCase(Node* node) {
        node->right = rightRotation(node->right);
        return rightRightCase(node);
    }

    Node* remove(Node* node, const T& elem) {
        if (node == nullptr) return TOKEN;

        int cmp = elem - node->value;

        if (cmp < 0) {
            Node* newLeftNode = remove(node->left, elem);
            if (newLeftNode == TOKEN) return TOKEN;
            node->left = newLeftNode;
        } else if (cmp > 0) {
            Node* newRightNode = remove(node->right, elem);
            if (newRightNode == TOKEN) return TOKEN;
            node->right = newRightNode;
        } else {
            if (node->left == nullptr) {
                return node->right;
            } else if (node->right == nullptr) {
                return node->left;
            } else {
                T successorValue = findMin(node->right);
                node->value = successorValue;
                Node* replacement = remove(node->right, successorValue);
                if (replacement == TOKEN) return TOKEN;
                node->right = replacement;
            }
        }

        update(node);
        return balance(node);
    }

    T findMin(Node* node) {
        while (node->left != nullptr) node = node->left;
        return node->value;
    }

    T findMax(Node* node) {
        while (node->right != nullptr) node = node->right;
        return node->value;
    }

    void destroy(Node* node) {
        if (node != nullptr) {
            destroy(node->left);
            destroy(node->right);
            delete node;
        }
    }

public:
    AVLTreeRecursiveOptimized() : root(nullptr), nodeCount(0), TOKEN(new Node(T())) {}

    ~AVLTreeRecursiveOptimized() {
        destroy(root);
        delete TOKEN;
    }

    int height() const {
        return (root == nullptr) ? 0 : root->height;
    }

    int size() const {
        return nodeCount;
    }

    bool isEmpty() const {
        return size() == 0;
    }

    bool contains(const T& value) const {
        Node* node = root;
        while (node != nullptr) {
            int cmp = value - node->value;
            if (cmp < 0) {
                node = node->left;
            } else if (cmp > 0) {
                node = node->right;
            } else {
                return true;
            }
        }
        return false;
    }

    bool insert(const T& value) {
        if (contains(value)) return false;
        Node* newRoot = insert(root, value);
        bool insertedNode = (newRoot != TOKEN);
        if (insertedNode) {
            nodeCount++;
            root = newRoot;
        }
        return insertedNode;
    }

    bool remove(const T& value) {
        Node* newRoot = remove(root, value);
        bool removedNode = (newRoot != TOKEN) || (newRoot == nullptr);

        if (removedNode) {
            root = newRoot;
            nodeCount--;
            return true;
        }

        return false;
    }

    class Iterator {
    private:
        std::stack<Node*> stack;
        Node* trav;
        int expectedNodeCount;

    public:
        Iterator(Node* root, int count) : trav(root), expectedNodeCount(count) {
            while (trav != nullptr) {
                stack.push(trav);
                trav = trav->left;
            }
        }

        bool hasNext() const {
            return trav != nullptr || !stack.empty();
        }

        T next() {
            if (expectedNodeCount != nodeCount) {
                throw std::runtime_error("Concurrent Modification Exception");
            }
            if (!hasNext()) {
                throw std::runtime_error("No such element");
            }
            Node* node = stack.top();
            stack.pop();
            trav = node->right;
            while (trav != nullptr) {
                stack.push(trav);
                trav = trav->left;
            }
            return node->value;
        }
    };

    Iterator begin() const {
        return Iterator(root, nodeCount);
    }

    Iterator end() const {
        return Iterator(nullptr, nodeCount);
    }

    bool validateBSTInvarient(Node* node) const {
        if (node == nullptr) return true;
        bool isValid = true;
        if (node->left != nullptr) isValid = isValid && node->left->value < node->value;
        if (node->right != nullptr) isValid = isValid && node->right->value > node->value;
        return isValid && validateBSTInvarient(node->left) && validateBSTInvarient(node->right);
    }
};

File: code/data_structures/balanced_binary_search_tree/AVLTreeRecursiveOptimized.py


class AVLTreeRecursiveOptimized:
    class Node:
        def __init__(self, value):
            self.bf = 0
            self.value = value
            self.height = 0
            self.left = None
            self.right = None

    def __init__(self):
        self.root = None
        self.nodeCount = 0
        self.TOKEN = self.Node(None)

    def height(self):
        if self.root is None:
            return 0
        return self.root.height

    def size(self):
        return self.nodeCount

    def is_empty(self):
        return self.size() == 0

    def contains(self, value):
        node = self.root
        while node is not None:
            cmp = value - node.value
            if cmp < 0:
                node = node.left
            elif cmp > 0:
                node = node.right
            else:
                return True
        return False

    def insert(self, value):
        if value is None:
            return False
        new_root = self._insert(self.root, value)
        inserted_node = new_root is not self.TOKEN
        if inserted_node:
            self.nodeCount += 1
            self.root = new_root
        return inserted_node

    def _insert(self, node, value):
        if node is None:
            return self.Node(value)
        cmp = value - node.value
        if cmp < 0:
            new_left_node = self._insert(node.left, value)
            if new_left_node is self.TOKEN:
                return self.TOKEN
            node.left = new_left_node
        elif cmp > 0:
            new_right_node = self._insert(node.right, value)
            if new_right_node is self.TOKEN:
                return self.TOKEN
            node.right = new_right_node
        else:
            return self.TOKEN
        self._update(node)
        return self._balance(node)

    def _update(self, node):
        left_node_height = -1 if node.left is None else node.left.height
        right_node_height = -1 if node.right is None else node.right.height
        node.height = 1 + max(left_node_height, right_node_height)
        node.bf = right_node_height - left_node_height

    def _balance(self, node):
        if node.bf == -2:
            if node.left.bf <= 0:
                return self.right_rotation(node)
            else:
                return self.left_right_case(node)
        elif node.bf == 2:
            if node.right.bf >= 0:
                return self.left_rotation(node)
            else:
                return self.right_left_case(node)
        return node

    def left_left_case(self, node):
        return self.right_rotation(node)

    def left_right_case(self, node):
        node.left = self.left_rotation(node.left)
        return self.left_left_case(node)

    def right_right_case(self, node):
        return self.left_rotation(node)

    def right_left_case(self, node):
        node.right = self.right_rotation(node.right)
        return self.right_right_case(node)

    def left_rotation(self, node):
        new_parent = node.right
        node.right = new_parent.left
        new_parent.left = node
        self._update(node)
        self._update(new_parent)
        return new_parent

    def right_rotation(self, node):
        new_parent = node.left
        node.left = new_parent.right
        new_parent.right = node
        self._update(node)
        self._update(new_parent)
        return new_parent

    def remove(self, elem):
        new_root = self._remove(self.root, elem)
        removed_node = new_root is not self.TOKEN or new_root is None
        if removed_node:
            self.root = new_root
            self.nodeCount -= 1
            return True
        return False

    def _remove(self, node, elem):
        if node is None:
            return self.TOKEN
        cmp = elem - node.value
        if cmp < 0:
            new_left_node = self._remove(node.left, elem)
            if new_left_node is self.TOKEN:
                return self.TOKEN
            node.left = new_left_node
        elif cmp > 0:
            new_right_node = self._remove(node.right, elem)
            if new_right_node is self.TOKEN:
                return self.TOKEN
            node.right = new_right_node
        else:
            if node.left is None:
                return node.right
            elif node.right is None:
                return node.left
            else:
                if node.left.height > node.right.height:
                    successor_value = self._find_max(node.left)
                    node.value = successor_value
                    replacement = self._remove(node.left, successor_value)
                    if replacement is self.TOKEN:
                        return self.TOKEN
                    node.left = replacement
                else:
                    successor_value = self._find_min(node.right)
                    node.value = successor_value
                    replacement = self._remove(node.right, successor_value)
                    if replacement is self.TOKEN:
                        return self.TOKEN
                    node.right = replacement
        self._update(node)
        return self._balance(node)

    def _find_min(self, node):
        while node.left is not None:
            node = node.left
        return node.value

    def _find_max(self, node):
        while node.right is not None:
            node = node.right
        return node.value

    def __iter__(self):
        expected_node_count = self.nodeCount
        stack = [self.root]

        def next_node():
            nonlocal stack
            nonlocal expected_node_count
            while stack:
                current_node = stack.pop()
                if current_node is None:
                    continue
                if current_node.right is not None:
                    stack.append(current_node.right)
                if current_node.left is not None:
                    stack.append(current_node.left)
                return current_node.value
            if expected_node_count != self.nodeCount:
                raise RuntimeError("Tree modified during iteration")
            raise StopIteration()

        return iter(next_node, None)

    def validate_BST_invariant(self, node):
        if node is None:
            return True
        val = node.value
        is_valid = True
        if node.left is not None:
            is_valid = is_valid and node.left.value < val
        if node.right is not None:
            is_valid = is_valid and node.right.value > val
        return is_valid and self.validate_BST_invariant(node.left) and self.validate_BST_invariant(node.right)
